### ISSUE TITLE
Value changed in a method isn't considered possible

### ISSUE THREAD

**Issue by nedbat** [microsoft/pyright#11738]
**Describe the bug**

A value at one point in the code is changed in a called method, but pyright thinks the value hasn't changed:
```python
    def arg_list(self) -> list[Ast]:
        args: list[Ast] = []
        if self.tok.kind == "lparen":  ## original value
            self.eat()
            while True:
                args.append(self.expr())
                if self.tok.kind == "rparen":   ## <==== error here
                    break
                self.eat("comma")
            self.eat("rparen")
        return args
```
The marked line is flagged with:
```
error: Condition will always evaluate to False since the types "Literal['lparen']" and "Literal['rparen']" have no overlap (reportUnnecessaryComparison)
```
pyright believes that `self.tok.kind` must still be "lparen" even though we've called two different methods that could have changed it (and did).

I tried making a smaller reproducer, but couldn't. This is in commit add0da7ddde91c1bd013eaafe325e5dcf96141ef in https://github.com/nedbat/acidica, using pyright 1.1.414. At that commit, this error is in parser.py, line 276.

**Comment by rchiodo** [comment]
Thanks for the report. This is a false positive for valid Python code, but it follows an intentional limitation of Pyright's type narrowing rather than a parser-specific bug.

### Why this happens

Pyright narrows expressions locally within the current execution scope:

1. `if self.tok.kind == "lparen"` narrows `self.tok.kind` from `str` to `Literal["lparen"]`.
2. `self.eat()` replaces `self.tok`, but Pyright does not propagate that method's mutations back into the caller's narrowing state. The same limitation applies to mutations performed by `self.expr()`.
3. The later comparison therefore appears to compare `Literal["lparen"]` against `"rparen"`, producing `reportUnnecessaryComparison`.

An explicit assignment to `self.tok` inside `arg_list` would invalidate the earlier narrowing of `self.tok.kind`. An assignment inside a called method does not.

This is the same limitation discussed in [#9442](https://github.com/microsoft/pyright/issues/9442#issuecomment-2468952824), which was classified as **as designed**, and in [this comment on #10311](https://github.com/microsoft/pyright/issues/10311#issuecomment-2800510336).

### Why calls do not simply discard narrowing

Consider:

```python
if self.value is not None:
    self.log("processing")
    self.value.process()
```

If every call invalidated attribute narrowing, the last line would require another `None` check even when `log` does not modify `value`. This would substantially reduce the usefulness of narrowing. Precisely identifying which calls can mutate which expressions instead requires interprocedural analysis that accounts for aliases, overrides, callbacks, and unavailable implementations. That would be a substantial design change, not a small fix for this comparison.

### Workaround

For a stateful parser, an ordinary Boolean predicate is a practical way to avoid establishing a narrowing that will become stale:

```python
def at(self, kind: str) -> bool:
    return self.tok.kind == kind
```

Use `if self.at("lparen"):` for the outer condition, and optionally `if self.at("rparen"):` for the inner one. An ordinary `bool`-returning method does not narrow `self.tok.kind` in its caller, so the opening-token check no longer establishes a stale literal type. This should be an ordinary Boolean predicate, not a `TypeGuard` or `TypeIs` function.

Importantly, change the **outer check**, where the stale narrowing originates, rather than just the line reporting the error.

Mutating parser state is perfectly reasonable; this limitation does not mean the parser is incorrect. Pyright simply does not model those mutations across method calls. Closing as **as designed** to reflect the existing analysis model.
